contrib/dyn_templates/src/lib.rs

```diff
@@ -439,3 +459,70 @@ impl Sentinel for Template {
         false
     }
 }
+
+/// A macro to easily create a template rendering context.
+///
+/// # Example
+///
+/// This code:
+///
+/// ```rust
+/// # #[macro_use] extern crate rocket;
+/// # use rocket_dyn_templates::{Template, context};
+///
+/// #[get("/<param>")]
+/// fn render_index(param: u64) -> Template {
+///     Template::render("index", context!{
+///         foo: "Hello world",
+///         bar: param,
+///     })
+/// }
+/// ```
+///
+/// is equivalent to the following, but without the need to manually define an
+/// `IndexContext` struct:
+///
+/// ```rust
+/// # #[macro_use] extern crate rocket;
+/// # use rocket_dyn_templates::Template;
+/// # use serde::Serialize;
+/// #[derive(Serialize)]
+/// struct IndexContext<'a> {
+///     foo: &'a str,
+///     bar: u64,
+/// }
+///
+/// #[get("/<param>")]
+/// fn render_index(param: u64) -> Template {
+///     Template::render("index", IndexContext {
+///         foo: "Hello world",
+///         bar: param,
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! context {
+    ($($key:ident $(: $value:expr)?),*$(,)?) => {{
+        use $crate::serde_internal::ser::{Serialize, Serializer, SerializeMap};
+
+        #[allow(non_camel_case_types)]
+        struct AnonymousInnerContextStruct<$($key: Serialize),*> {
+            $($key: $key),*
+        }
+
+        #[allow(non_camel_case_types)]
+        impl<$($key: Serialize),*> Serialize for AnonymousInnerContextStruct<$($key),*> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where S: Serializer,
+            {
+                let mut map = serializer.serialize_map(None)?;
+                $(map.serialize_entry(stringify!($key), &self.$key)?;)*
+                map.end()
+            }
+        }
+
+        AnonymousInnerContextStruct {
+            $($key $(: &$value)?),*
```

Comment:
Interesting, not to take ownership. I'm not sure if this is what we want. Can you argue one direction or another? The only argument I have for taking ownership is that not doing so might be surprising.￼

Perhaps more concerning is that ownership is taken when shorthand syntax is used but not when the full syntax is used.￼

Reply:
My reasoning for not taking ownership was just following in line with e.g. `println!`. Plus, taking ownership when it's not needed seems unnecessarily limiting. I agree that the differing behavior with the shorthand syntax is bad, I didn't catch that one earlier.

I'm all for borrowing (& fixing the shorthand syntax) unless you think ownership is the better path?

Reply:
I think ownership is the more obvious path. `println!` doesn't generate a value, so its approach in sensible there. Given we _do_ generate a value, I don't think borrowing by default makes sense. It would make it possible, for instance, to do something like:

```rust
fn f() -> impl Serialize {
    let foo = String::new();
    context! { foo }
}
```

If we want to get creative, we can introduce some new shorthand syntax (which I've quickly grown fond of):

```rust
let foo = String::new();
context! { ref foo } == context! { foo: &foo }
```

Or just leave it as is.

Reply:
Good point, makes sense. I just changed the macro to take ownership of fields.
I think we should move forward with the syntax as-is for a MVP, and we can possibly add `ref` later on after we get a feel for how the macro is being used.